# Concepts

#### We have Books, Rating and User csv file, from where we will pick up book and rating file.

#### Since the book have book name and ratings have user id, book id and rating, we'll make a new data frame that will make n users as its column and book name will be in row of the data frame. Value of the each row will be rate done by a particular user.

#### Here we won't be considering all of the books and users. We'll consider only those users who have given rating to more than 200 books and we'll choose those books whose total rating is greater than 50.

####

####


In [4]:
# importing library
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [5]:
# importing the df
books = pd.read_csv("../csv/Books.csv")
ratings = pd.read_csv("../csv/Ratings.csv")
users = pd.read_csv("../csv/Users.csv")

/tmp/ipykernel_6605/4069772381.py:2: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  books = pd.read_csv("../csv/Books.csv")


### **Merging rating dataframe with books**

In [6]:
# Merging the book with rating df
rating_with_name = ratings.merge(books, on="ISBN")

In [7]:
print(f"rating shape:{ratings.shape}\nbooks shape:{books.shape}\nusers shape: {users.shape}\nrating with name shape:{rating_with_name.shape}")

rating shape:(1149780, 3)
books shape:(271360, 8)
users shape: (278858, 3)
rating with name shape:(1031136, 10)


In [8]:
rating_with_name.head()

User-ID        ISBN  Book-Rating  \
0   276725  034545104X            0   
1   276726  0155061224            5   
2   276727  0446520802            0   
3   276729  052165615X            3   
4   276729  0521795028            6   

                                          Book-Title      Book-Author  \
0                               Flesh Tones: A Novel       M. J. Rose   
1                                   Rites of Passage       Judith Rae   
2                                       The Notebook  Nicholas Sparks   
3                                     Help!: Level 1    Philip Prowse   
4  The Amsterdam Connection : Level 4 (Cambridge ...      Sue Leather   

  Year-Of-Publication                   Publisher  \
0                2002            Ballantine Books   
1                2001                      Heinle   
2                1996                Warner Books   
3                1999  Cambridge University Press   
4                2001  Cambridge University Press   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/034545104X.0...   
1  http://images.amazon.com/images/P/0155061224.0...   
2  http://images.amazon.com/images/P/0446520802.0...   
3  http://images.amazon.com/images/P/052165615X.0...   
4  http://images.amazon.com/images/P/0521795028.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/034545104X.0...   
1  http://images.amazon.com/images/P/0155061224.0...   
2  http://images.amazon.com/images/P/0446520802.0...   
3  http://images.amazon.com/images/P/052165615X.0...   
4  http://images.amazon.com/images/P/0521795028.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/034545104X.0...  
1  http://images.amazon.com/images/P/0155061224.0...  
2  http://images.amazon.com/images/P/0446520802.0...  
3  http://images.amazon.com/images/P/052165615X.0...  
4  http://images.amazon.com/images/P/0521795028.0...

### We're filtering out such that only those users who have done minimum of 200 votings will be kept in df.

In [9]:
# groupby function in pandas is used to split the data into groups based on some criteria
"""This groups the DataFrame by the "User-ID" column. Each unique value in "User-ID" becomes a group, and all rows with the same "User-ID" are placed into that group. After grouping by "User-ID", the count() function is applied. This function counts the number of non-NA/null entries in each column for each group. Finally the ['Book-Rating'] helps to retreive only the Book-Rating column"""
voted_users = rating_with_name.groupby("User-ID").count()["Book-Rating"]

# doing boolean indexing: here voted_user return the record of all the user who have voted for 1 or more time.
# voted_user>200 returns boolean of either True(if his/her vote count is >200) or false(if his/her vote count is <200)
# finally we did boolean indexing in the voted_user var which return those users record who have voted for more than 200 time.
experience_voted_user = voted_users[voted_users > 200]
# padhe_likhe_users
experience_voted_user_index = experience_voted_user.index
experience_voted_user
# It shows only 811 unique users have voted for more than 200 times

User-ID
254        300
2276       456
2766       269
2977       227
3363       890
          ... 
274308    1293
275970    1325
277427     490
277639     265
278418    3996
Name: Book-Rating, Length: 811, dtype: int64

In [10]:
experience_voted_user_index

Index([   254,   2276,   2766,   2977,   3363,   4017,   4385,   6251,   6323,
         6543,
       ...
       271705, 273979, 274004, 274061, 274301, 274308, 275970, 277427, 277639,
       278418],
      dtype='int64', name='User-ID', length=811)

- The length of above result shows out of 278858 users in our data set only 92106 have rated books
- Again from these result of user voted, we will only choose those use who have voted for more than 200 times 

### Now that we've got the User-ID of those user who have voted for more than 200 times, we will filter out from rating_with_name dataframe to grab those books which have obtain voting from more than 50 readers

In [11]:
filtered_rating = rating_with_name[
    rating_with_name["User-ID"].isin(experience_voted_user_index)
]

In [12]:
filtered_rating

User-ID        ISBN  Book-Rating  \
1150      277427  002542730X           10   
1151      277427  0026217457            0   
1152      277427  003008685X            8   
1153      277427  0030615321            0   
1154      277427  0060002050            0   
...          ...         ...          ...   
1029357   275970  1931868123            0   
1029358   275970  3411086211           10   
1029359   275970  3829021860            0   
1029360   275970  4770019572            0   
1029361   275970  9626340762            8   

                                                Book-Title  \
1150     Politically Correct Bedtime Stories: Modern Ta...   
1151                    Vegetarian Times Complete Cookbook   
1152                                              Pioneers   
1153      Ask for May, Settle for June (A Doonesbury book)   
1154                     On a Wicked Dawn (Cynster Novels)   
...                                                    ...   
1029357  There's a Porcupine in My Outhouse: Misadventu...   
1029358                                         Die Biene.   
1029359                                     The Penis Book   
1029360                                            Musashi   
1029361  Northanger Abbey (Classic Literature with Clas...   

                           Book-Author Year-Of-Publication  \
1150                 James Finn Garner                1994   
1151                        Lucy  Moll                1995   
1152             James Fenimore Cooper                1974   
1153                     G. B. Trudeau                1982   
1154                 Stephanie Laurens                2002   
...                                ...                 ...   
1029357                   Mike Tougias                2002   
1029358  Sybil Gr�¤fin Sch�¶nfeldt                1993   
1029359                   Joseph Cohen                1999   
1029360                 Eiji Yoshikawa                1995   
1029361                    Jane Austen                1996   

                                    Publisher  \
1150                John Wiley &amp; Sons Inc   
1151                    John Wiley &amp; Sons   
1152                         Thomson Learning   
1153                      Henry Holt &amp; Co   
1154                               Avon Books   
...                                       ...   
1029357                    Capital Books (VA)   
1029358  Bibliographisches Institut, Mannheim   
1029359                              Konemann   
1029360          Kodansha International (JPN)   
1029361                 Naxos Audiobooks Ltd.   

                                               Image-URL-S  \
1150     http://images.amazon.com/images/P/002542730X.0...   
1151     http://images.amazon.com/images/P/0026217457.0...   
1152     http://images.amazon.com/images/P/003008685X.0...   
1153     http://images.amazon.com/images/P/0030615321.0...   
1154     http://images.amazon.com/images/P/0060002050.0...   
...                                                    ...   
1029357  http://images.amazon.com/images/P/1931868123.0...   
1029358  http://images.amazon.com/images/P/3411086211.0...   
1029359  http://images.amazon.com/images/P/3829021860.0...   
1029360  http://images.amazon.com/images/P/4770019572.0...   
1029361  http://images.amazon.com/images/P/9626340762.0...   

                                               Image-URL-M  \
1150     http://images.amazon.com/images/P/002542730X.0...   
1151     http://images.amazon.com/images/P/0026217457.0...   
1152     http://images.amazon.com/images/P/003008685X.0...   
1153     http://images.amazon.com/images/P/0030615321.0...   
1154     http://images.amazon.com/images/P/0060002050.0...   
...                                                    ...   
1029357  http://images.amazon.com/images/P/1931868123.0...   
1029358  http://images.amazon.com/images/P/3411086211.0...   
1029359  http://images.amazon.com/images/P/3829021860.0...   
1029360  http://images.a

### From the obtained DF, we will filter out those books whose rating is less than 50 and keep only those whose rating is greater than 50

In [13]:
books_with_rating = filtered_rating.groupby("Book-Title").count()["Book-Rating"]
books_with_rating

Book-Title
 A Light in the Storm: The Civil War Diary of Amelia Martin, Fenwick Island, Delaware, 1861 (Dear America)    2
 Always Have Popsicles                                                                                        1
 Apple Magic (The Collector's series)                                                                         1
 Beyond IBM: Leadership Marketing and Finance for the 1990s                                                   1
 Clifford Visita El Hospital (Clifford El Gran Perro Colorado)                                                1
                                                                                                             ..
Ã?Â?ber das Fernsehen.                                                                                        2
Ã?Â?ber die Pflicht zum Ungehorsam gegen den Staat.                                                           3
Ã?Â?lpiraten.                                                                                

In [14]:
# Performing boolean indexing which gives us the book with rating greater than 50
books_with_higher_rating = books_with_rating[books_with_rating >= 50]
index_books_with_higher_rating = books_with_higher_rating.index
books_with_higher_rating

Book-Title
1984                                                                  70
1st to Die: A Novel                                                  160
2nd Chance                                                           122
4 Blondes                                                             70
A Bend in the Road                                                   114
                                                                    ... 
Year of Wonders                                                       57
You Belong To Me                                                      55
Zen and the Art of Motorcycle Maintenance: An Inquiry into Values     62
Zoya                                                                  59
\O\" Is for Outlaw"                                                  101
Name: Book-Rating, Length: 706, dtype: int64

In [15]:
index_books_with_higher_rating

Index(['1984', '1st to Die: A Novel', '2nd Chance', '4 Blondes',
       'A Bend in the Road', 'A Case of Need',
       'A Child Called \It\": One Child's Courage to Survive"',
       'A Civil Action', 'A Day Late and a Dollar Short', 'A Fine Balance',
       ...
       'Winter Solstice', 'Wish You Well', 'Without Remorse',
       'Wizard and Glass (The Dark Tower, Book 4)', 'Wuthering Heights',
       'Year of Wonders', 'You Belong To Me',
       'Zen and the Art of Motorcycle Maintenance: An Inquiry into Values',
       'Zoya', '\O\" Is for Outlaw"'],
      dtype='object', name='Book-Title', length=706)

### This is the final filtered dataframe which contains those user whose voting_count>200 and those books whos rating_count>50

In [16]:
final_filtered_rating = filtered_rating[
    filtered_rating["Book-Title"].isin(index_books_with_higher_rating)
]
final_filtered_rating

User-ID        ISBN  Book-Rating  \
1150      277427  002542730X           10   
1163      277427  0060930535            0   
1165      277427  0060934417            0   
1168      277427  0061009059            9   
1174      277427  006440188X            0   
...          ...         ...          ...   
1029196   275970  1400031354            0   
1029197   275970  1400031362            0   
1029270   275970  1573229725            0   
1029309   275970  1586210661            9   
1029310   275970  1586212230            0   

                                                Book-Title  \
1150     Politically Correct Bedtime Stories: Modern Ta...   
1163                         The Poisonwood Bible: A Novel   
1165                                    Bel Canto: A Novel   
1168     One for the Money (Stephanie Plum Novels (Pape...   
1174                                     The Secret Garden   
...                                                    ...   
1029196  Tears of the Giraffe (No.1 Ladies Detective Ag...   
1029197  Morality for Beautiful Girls (No.1 Ladies Dete...   
1029270                                        Fingersmith   
1029309                             Me Talk Pretty One Day   
1029310                                              Naked   

                     Book-Author Year-Of-Publication  \
1150           James Finn Garner                1994   
1163          Barbara Kingsolver                1999   
1165                Ann Patchett                2002   
1168             Janet Evanovich                1995   
1174     Frances Hodgson Burnett                1998   
...                          ...                 ...   
1029196   Alexander McCall Smith                2002   
1029197   Alexander McCall Smith                2002   
1029270             Sarah Waters                2002   
1029309            David Sedaris                2001   
1029310            David Sedaris                2001   

                         Publisher  \
1150     John Wiley &amp; Sons Inc   
1163                     Perennial   
1165                     Perennial   
1168                   HarperTorch   
1174                  HarperTrophy   
...                            ...   
1029196                     Anchor   
1029197                     Anchor   
1029270            Riverhead Books   
1029309    Time Warner Audio Major   
1029310    Time Warner Audio Major   

                                               Image-URL-S  \
1150     http://images.amazon.com/images/P/002542730X.0...   
1163     http://images.amazon.com/images/P/0060930535.0...   
1165     http://images.amazon.com/images/P/0060934417.0...   
1168     http://images.amazon.com/images/P/0061009059.0...   
1174     http://images.amazon.com/images/P/006440188X.0...   
...                                                    ...   
1029196  http://images.amazon.com/images/P/1400031354.0...   
1029197  http://images.amazon.com/images/P/1400031362.0...   
1029270  http://images.amazon.com/images/P/1573229725.0...   
1029309  http://images.amazon.com/images/P/1586210661.0...   
1029310  http://images.amazon.com/images/P/1586212230.0...   

                                               Image-URL-M  \
1150     http://images.amazon.com/images/P/002542730X.0...   
1163     http://images.amazon.com/images/P/0060930535.0...   
1165     http://images.amazon.com/images/P/0060934417.0...   
1168     http://images.amazon.com/images/P/0061009059.0...   
1174     http://images.amazon.com/images/P/006440188X.0...   
...                                                    ...   
1029196  http://images.amazon.com/images/P/1400031354.0...   
1029197  http://images.amazon.com/images/P/1400031362.0...   
1029270  http://images.amazon.com/images/P/1573229725.0...   
1029309  http://images.amazon.com/images/P/1586210661.0...   
1029310  http://images.amazon.com/images/P/1586212230.0...   

                                               Image-URL-L  
1150     http://images.amazon.com/images/P

In [18]:
final_filtered_rating1 = final_filtered_rating
# checking whether there is duplcates or not
final_filtered_rating1.drop_duplicates().shape

(58586, 10)

## As we have got the final filtered dataframe, we gonna use pivot table to creata new dataframe where in column there will be user id and in row there will be book isbn

In [62]:
pt = final_filtered_rating.pivot_table(
    index="Book-Title", columns="User-ID", values="Book-Rating"
)

In [64]:
pt

User-ID                                             254     2276    2766    \
Book-Title                                                                   
1984                                                   9.0     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN    10.0     NaN   
4 Blondes                                              NaN     NaN     NaN   
A Bend in the Road                                     0.0     NaN     7.0   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     NaN     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             2977    3363    4017    \
Book-Title                                                                   
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     NaN     NaN   
A Bend in the Road                                     NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     0.0     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             4385    6251    6323    \
Book-Title                                                                   
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     0.0     NaN   
A Bend in the Road                                     NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     NaN   
You Belong To Me                                       NaN     NaN     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     0.0     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             6543    ...  271705  \
Book-Title                                                  ...           
1984                                                   NaN  ...    10.0   
1st to Die: A Novel                                    9.0  ...     NaN   
2nd Chance                                             0.0  ...     NaN   
4 Blondes                                              NaN  ...     NaN   
A Bend in the Road                                     NaN  ...     NaN   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     NaN   
You Belong To Me                                       NaN  ...     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN  ...     NaN   
Zoya                                                   NaN  ...     NaN   
\O\" Is for Outlaw"                                    

In [66]:
# replacing the NaN value with 0
pt.fillna(0, inplace=True)
pt

User-ID                                             254     2276    2766    \
Book-Title                                                                   
1984                                                   9.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0    10.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     7.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             2977    3363    4017    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             4385    6251    6323    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             6543    ...  271705  \
Book-Title                                                  ...           
1984                                                   0.0  ...    10.0   
1st to Die: A Novel                                    9.0  ...     0.0   
2nd Chance                                             0.0  ...     0.0   
4 Blondes                                              0.0  ...     0.0   
A Bend in the Road                                     0.0  ...     0.0   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     0.0   
You Belong To Me                                       0.0  ...     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     0.0   
Zoya                                                   0.0  ...     0.0   
\O\" Is for Outlaw"                                    